GVFS/GVFS.Common/ProductUpgraderInfo.cs

```diff
@@ -10,39 +10,59 @@ public class ProductUpgraderInfo
         public const string UpgradeDirectoryName = "GVFS.Upgrade";
         public const string LogDirectory = "Logs";
         public const string DownloadDirectory = "Downloads";
+        public const string HighestAvailableVersionFileName = "HighestAvailableVersion";
 
         protected const string RootDirectory = UpgradeDirectoryName;
 
-        private static readonly HashSet<string> GVFSInstallerFileNamePrefixCandidates = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+        public static bool IsLocalUpgradeAvailable(ITracer tracer)
         {
-            "SetupGVFS",
-            "VFSForGit"
-        };
+            try
+            {
+                Version highestAvailableVersion = ReadHighestRecordedAvailableVersion();
+                if (highestAvailableVersion == null)
+                {
+                    return false;
+                }
 
-        public static bool IsLocalUpgradeAvailable(string installerExtension)
-        {
-            string downloadDirectory = GetAssetDownloadsPath();
-            if (Directory.Exists(downloadDirectory))
+                Version currentVersion = new Version(CurrentGVFSVersion());
+                return highestAvailableVersion > currentVersion;

```

Comment:
We previously weren't comparing the downloaded versions against our own process version.  Would it be enough to just test for the existence of a marker file here?

Reply:
Probably - the benefit of reading the version here is we can confirm that we are out-of-date.

The window of when this file would exist, but the user is actually up-to-date is pretty small - in particular it could happen when a user manually installs a new version (not via upgrade), but the background process has not cleared the file yet.

That brings up a point I missed - upgrade should clear this file when it runs... Although as is, the user would not see incorrect results due to this check.

Reply:
> Probably - the benefit of reading the version here is we can confirm that we are out-of-date.

That makes sense.  I wasn't sure how much extra overhead we're adding to the hooks with  open+read+parse vs a file existence check.